File: ios/JustSpentTests/CurrencyTests.swift
//
//  CurrencyTests.swift
//  JustSpentTests
//
//  Created by Claude Code on 2025-10-19.
//  Comprehensive tests for currency functionality
//

import XCTest
@testable import JustSpent

final class CurrencyTests: XCTestCase {

    // MARK: - Currency Model Tests

    func testCurrencyProperties() {
        // Test USD
        let usd = Currency.usd
        XCTAssertEqual(usd.rawValue, "USD")
        XCTAssertEqual(usd.symbol, "$")
        XCTAssertEqual(usd.displayName, "US Dollar")
        XCTAssertFalse(usd.isRTL)

        // Test AED
        let aed = Currency.aed
        XCTAssertEqual(aed.rawValue, "AED")
        XCTAssertEqual(aed.symbol, "د.إ")
        XCTAssertEqual(aed.displayName, "UAE Dirham")
        XCTAssertTrue(aed.isRTL)

        // Test EUR
        let eur = Currency.eur
        XCTAssertEqual(eur.rawValue, "EUR")
        XCTAssertEqual(eur.symbol, "€")
        XCTAssertEqual(eur.displayName, "Euro")
        XCTAssertFalse(eur.isRTL)
    }

    func testCurrencyFromISOCode() {
        XCTAssertEqual(Currency.from(isoCode: "USD"), .usd)
        XCTAssertEqual(Currency.from(isoCode: "AED"), .aed)
        XCTAssertEqual(Currency.from(isoCode: "EUR"), .eur)
        XCTAssertEqual(Currency.from(isoCode: "GBP"), .gbp)
        XCTAssertEqual(Currency.from(isoCode: "INR"), .inr)
        XCTAssertEqual(Currency.from(isoCode: "SAR"), .sar)

        // Test case insensitivity
        XCTAssertEqual(Currency.from(isoCode: "usd"), .usd)
        XCTAssertEqual(Currency.from(isoCode: "Aed"), .aed)

        // Test invalid code
        XCTAssertNil(Currency.from(isoCode: "INVALID"))
    }

    func testCurrencyDetectionFromText() {
        // Test symbol detection
        XCTAssertEqual(Currency.detectFromText("$50"), .usd)
        XCTAssertEqual(Currency.detectFromText("د.إ 100"), .aed)
        XCTAssertEqual(Currency.detectFromText("€25"), .eur)
        XCTAssertEqual(Currency.detectFromText("£20"), .gbp)
        XCTAssertEqual(Currency.detectFromText("₹500"), .inr)
        XCTAssertEqual(Currency.detectFromText("﷼50"), .sar)

        // Test name detection
        XCTAssertEqual(Currency.detectFromText("50 dollars"), .usd)
        XCTAssertEqual(Currency.detectFromText("100 dirhams"), .aed)
        XCTAssertEqual(Currency.detectFromText("25 euros"), .eur)
        XCTAssertEqual(Currency.detectFromText("20 pounds"), .gbp)
        XCTAssertEqual(Currency.detectFromText("500 rupees"), .inr)
        XCTAssertEqual(Currency.detectFromText("50 riyals"), .sar)

        // Test colloquial terms
        XCTAssertEqual(Currency.detectFromText("50 bucks"), .usd)
        XCTAssertEqual(Currency.detectFromText("20 quid"), .gbp)

        // Test no match
        XCTAssertNil(Currency.detectFromText("spent money"))
    }

    func testCurrencyVoiceKeywords() {
        let usd = Currency.usd
        XCTAssertTrue(usd.voiceKeywords.contains("usd"))
        XCTAssertTrue(usd.voiceKeywords.contains("dollar"))
        XCTAssertTrue(usd.voiceKeywords.contains("$"))

        let aed = Currency.aed
        XCTAssertTrue(aed.voiceKeywords.contains("aed"))
        XCTAssertTrue(aed.voiceKeywords.contains("dirham"))
        XCTAssertTrue(aed.voiceKeywords.contains("د.إ"))
    }

    // MARK: - Currency Formatter Tests

    func testBasicFormatting() {
        let amount: Decimal = 1234.56

        // USD
        let usdFormatted = CurrencyFormatter.shared.format(
            amount: amount,
            currency: .usd,
            showSymbol: true,
            showCode: false
        )
        XCTAssertTrue(usdFormatted.contains("$"))
        XCTAssertTrue(usdFormatted.contains("1,234.56") || usdFormatted.contains("1234.56"))

        // AED
        let aedFormatted = CurrencyFormatter.shared.format(
            amount: amount,
            currency: .aed,
            showSymbol: true,
            showCode: false
        )
        XCTAssertTrue(aedFormatted.contains("د.إ"))

        // EUR
        let eurFormatted = CurrencyFormatter.shared.format(
            amount: amount,
            currency: .eur,
            showSymbol: true,
            showCode: false
        )
        XCTAssertTrue(eurFormatted.contains("€"))
    }

    func testFormattingWithCode() {
        let amount: Decimal = 100.00

        let formatted = CurrencyFormatter.shared.format(
            amount: amount,
            currency: .usd,
            showSymbol: true,
            showCode: true
        )

        XCTAssertTrue(formatted.contains("$"))
        XCTAssertTrue(formatted.contains("USD"))
    }

    func testCompactFormatting() {
        let amount: Decimal = 50.25

        let compact = CurrencyFormatter.shared.formatCompact(amount: amount, currency: .usd)
        XCTAssertTrue(compact.contains("$"))
        XCTAssertTrue(compact.contains("50.25"))
    }

    func testDetailedFormatting() {
        let amount: Decimal = 75.50

        let detailed = CurrencyFormatter.shared.formatDetailed(amount: amount, currency: .gbp)
        XCTAssertTrue(detailed.contains("£"))
        XCTAssertTrue(detailed.contains("GBP"))
    }

    func testDecimalExtensions() {
        let amount: Decimal = 100.00

        let formatted = amount.formatted(as: .usd)
        XCTAssertTrue(formatted.contains("$"))

        let compact = amount.formattedCompact(as: .aed)
        XCTAssertTrue(compact.contains("د.إ"))

        let detailed = amount.formattedDetailed(as: .eur)
        XCTAssertTrue(detailed.contains("€"))
        XCTAssertTrue(detailed.contains("EUR"))
    }

    func testCurrencyParsing() {
        // Test parsing formatted currency
        let parsed = CurrencyFormatter.shared.parse(string: "$100.50", currency: .usd)
        XCTAssertEqual(parsed, Decimal(string: "100.50"))

        // Test parsing with different symbols
        let parsedAED = CurrencyFormatter.shared.parse(string: "د.إ 250.75", currency: .aed)
        XCTAssertEqual(parsedAED, Decimal(string: "250.75"))

        // Test parsing just numbers
        let parsedPlain = CurrencyFormatter.shared.parse(string: "50.25", currency: .usd)
        XCTAssertEqual(parsedPlain, Decimal(string: "50.25"))
    }

    // MARK: - Voice Currency Detector Tests

    func testSimpleAmountCurrencyExtraction() {
        let detector = VoiceCurrencyDetector.shared

        // Test "50 dollars"
        if let result = detector.extractAmountAndCurrency(from: "50 dollars") {
            XCTAssertEqual(result.amount, 50)
            XCTAssertEqual(result.currency, .usd)
        } else {
            XCTFail("Failed to extract amount and currency")
        }

        // Test "د.إ 100"
        if let result = detector.extractAmountAndCurrency(from: "د.إ 100") {
            XCTAssertEqual(result.amount, 100)
            XCTAssertEqual(result.currency, .aed)
        } else {
            XCTFail("Failed to extract amount and currency")
        }
    }

    func testComplexVoiceCommandExtraction() {
        let detector = VoiceCurrencyDetector.shared

        // Test full sentence
        if let result = detector.extractAmountAndCurrency(from: "I just spent 25.50 euros on groceries") {
            XCTAssertEqual(result.amount, 25.50)
            XCTAssertEqual(result.currency, .eur)
        } else {
            XCTFail("Failed to extract from complex sentence")
        }

        // Test with merchant
        if let result = detector.extractAmountAndCurrency(from: "Paid 20 pounds at the store") {
            XCTAssertEqual(result.amount, 20)
            XCTAssertEqual(result.currency, .gbp)
        } else {
            XCTFail("Failed to extract from sentence with merchant")
        }
    }

    func testCurrencyDetectionWithoutAmount() {
        let detector = VoiceCurrencyDetector.shared

        let usd = detector.detectCurrency(from: "spent some dollars", default: .aed)
        XCTAssertEqual(usd, .usd)

        let aed = detector.detectCurrency(from: "paid in dirhams", default: .usd)
        XCTAssertEqual(aed, .aed)

        // Test default fallback
        let defaultCurrency = detector.detectCurrency(from: "spent money", default: .gbp)
        XCTAssertEqual(defaultCurrency, .gbp)
    }

    func testContainsCurrency() {
        let detector = VoiceCurrencyDetector.shared

        XCTAssertTrue(detector.containsCurrency("50 dollars"))
        XCTAssertTrue(detector.containsCurrency("د.إ 100"))
        XCTAssertTrue(detector.containsCurrency("paid in euros"))
        XCTAssertFalse(detector.containsCurrency("spent money"))
    }

    func testNormalizeCurrencySymbols() {
        let detector = VoiceCurrencyDetector.shared

        let normalized = detector.normalizeCurrencySymbols(in: "$50 and €25 and د.إ 100")
        XCTAssertTrue(normalized.contains("USD"))
        XCTAssertTrue(normalized.contains("EUR"))
        XCTAssertTrue(normalized.contains("AED"))
    }

    // MARK: - User Preferences Tests

    func testDefaultCurrencyPersistence() {
        let preferences = UserPreferences.shared

        // Set new currency
        preferences.setDefaultCurrency(.aed)

        // Verify it's saved
        XCTAssertEqual(preferences.getCurrentCurrency(), .aed)

        // Reset to default
        preferences.resetToDefaults()
    }

    func testCurrencyChangeNotification() {
        let preferences = UserPreferences.shared
        let expectation = XCTestExpectation(description: "Currency changed")

        var receivedCurrency: Currency?
        let cancellable = preferences.$defaultCurrency
            .dropFirst() // Skip initial value
            .sink { currency in
                receivedCurrency = currency
                expectation.fulfill()
            }

        // Change currency
        preferences.setDefaultCurrency(.eur)

        wait(for: [expectation], timeout: 1.0)
        XCTAssertEqual(receivedCurrency, .eur)

        cancellable.cancel()
        preferences.resetToDefaults()
    }

    // MARK: - Integration Tests

    func testEndToEndCurrencyFlow() {
        // Simulate voice command: "I spent 50 dirhams on groceries"
        let voiceCommand = "I spent 50 dirhams on groceries"

        // 1. Detect currency
        guard let currency = Currency.detectFromText(voiceCommand) else {
            XCTFail("Failed to detect currency")
            return
        }
        XCTAssertEqual(currency, .aed)

        // 2. Extract amount
        let detector = VoiceCurrencyDetector.shared
        guard let (amount, extractedCurrency) = detector.extractAmountAndCurrency(from: voiceCommand) else {
            XCTFail("Failed to extract amount and currency")
            return
        }
        XCTAssertEqual(amount, 50)
        XCTAssertEqual(extractedCurrency, .aed)

        // 3. Format for display
        let formatted = CurrencyFormatter.shared.formatCompact(amount: amount, currency: extractedCurrency)
        XCTAssertTrue(formatted.contains("د.إ"))
        XCTAssertTrue(formatted.contains("50"))
    }
}

File: android/app/src/test/java/com/justspent/app/CurrencyTests.kt
package com.justspent.app

import com.justspent.app.data.model.Currency
import com.justspent.app.utils.CurrencyFormatter
import com.justspent.app.utils.VoiceCurrencyDetector
import com.justspent.app.utils.formatted
import com.justspent.app.utils.formattedCompact
import com.justspent.app.utils.formattedDetailed
import org.junit.Assert.*
import org.junit.Test
import java.math.BigDecimal

/**
 * Comprehensive tests for currency functionality
 */
class CurrencyTests {

    // MARK: - Currency Model Tests

    @Test
    fun testCurrencyProperties() {
        // Test USD
        val usd = Currency.USD
        assertEquals("USD", usd.code)
        assertEquals("$", usd.symbol)
        assertEquals("US Dollar", usd.displayName)
        assertFalse(usd.isRTL)

        // Test AED
        val aed = Currency.AED
        assertEquals("AED", aed.code)
        assertEquals("د.إ", aed.symbol)
        assertEquals("UAE Dirham", aed.displayName)
        assertTrue(aed.isRTL)

        // Test EUR
        val eur = Currency.EUR
        assertEquals("EUR", eur.code)
        assertEquals("€", eur.symbol)
        assertEquals("Euro", eur.displayName)
        assertFalse(eur.isRTL)
    }

    @Test
    fun testCurrencyFromCode() {
        assertEquals(Currency.USD, Currency.fromCode("USD"))
        assertEquals(Currency.AED, Currency.fromCode("AED"))
        assertEquals(Currency.EUR, Currency.fromCode("EUR"))
        assertEquals(Currency.GBP, Currency.fromCode("GBP"))
        assertEquals(Currency.INR, Currency.fromCode("INR"))
        assertEquals(Currency.SAR, Currency.fromCode("SAR"))

        // Test case insensitivity
        assertEquals(Currency.USD, Currency.fromCode("usd"))
        assertEquals(Currency.AED, Currency.fromCode("Aed"))

        // Test invalid code
        assertNull(Currency.fromCode("INVALID"))
    }

    @Test
    fun testCurrencyDetectionFromText() {
        // Test symbol detection
        assertEquals(Currency.USD, Currency.detectFromText("$50"))
        assertEquals(Currency.AED, Currency.detectFromText("د.إ 100"))
        assertEquals(Currency.EUR, Currency.detectFromText("€25"))
        assertEquals(Currency.GBP, Currency.detectFromText("£20"))
        assertEquals(Currency.INR, Currency.detectFromText("₹500"))
        assertEquals(Currency.SAR, Currency.detectFromText("﷼50"))

        // Test name detection
        assertEquals(Currency.USD, Currency.detectFromText("50 dollars"))
        assertEquals(Currency.AED, Currency.detectFromText("100 dirhams"))
        assertEquals(Currency.EUR, Currency.detectFromText("25 euros"))
        assertEquals(Currency.GBP, Currency.detectFromText("20 pounds"))
        assertEquals(Currency.INR, Currency.detectFromText("500 rupees"))
        assertEquals(Currency.SAR, Currency.detectFromText("50 riyals"))

        // Test colloquial terms
        assertEquals(Currency.USD, Currency.detectFromText("50 bucks"))
        assertEquals(Currency.GBP, Currency.detectFromText("20 quid"))

        // Test no match
        assertNull(Currency.detectFromText("spent money"))
    }

    @Test
    fun testCurrencyVoiceKeywords() {
        val usd = Currency.USD
        assertTrue(usd.voiceKeywords.contains("usd"))
        assertTrue(usd.voiceKeywords.contains("dollar"))
        assertTrue(usd.voiceKeywords.contains("$"))

        val aed = Currency.AED
        assertTrue(aed.voiceKeywords.contains("aed"))
        assertTrue(aed.voiceKeywords.contains("dirham"))
        assertTrue(aed.voiceKeywords.contains("د.إ"))
    }

    // MARK: - Currency Formatter Tests

    @Test
    fun testBasicFormatting() {
        val amount = BigDecimal("1234.56")

        // USD
        val usdFormatted = CurrencyFormatter.format(
            amount = amount,
            currency = Currency.USD,
            showSymbol = true,
            showCode = false
        )
        assertTrue(usdFormatted.contains("$"))
        assertTrue(usdFormatted.contains("1,234.56") || usdFormatted.contains("1234.56"))

        // AED
        val aedFormatted = CurrencyFormatter.format(
            amount = amount,
            currency = Currency.AED,
            showSymbol = true,
            showCode = false
        )
        assertTrue(aedFormatted.contains("د.إ"))

        // EUR
        val eurFormatted = CurrencyFormatter.format(
            amount = amount,
            currency = Currency.EUR,
            showSymbol = true,
            showCode = false
        )
        assertTrue(eurFormatted.contains("€"))
    }

    @Test
    fun testFormattingWithCode() {
        val amount = BigDecimal("100.00")

        val formatted = CurrencyFormatter.format(
            amount = amount,
            currency = Currency.USD,
            showSymbol = true,
            showCode = true
        )

        assertTrue(formatted.contains("$"))
        assertTrue(formatted.contains("USD"))
    }

    @Test
    fun testCompactFormatting() {
        val amount = BigDecimal("50.25")

        val compact = CurrencyFormatter.formatCompact(amount, Currency.USD)
        assertTrue(compact.contains("$"))
        assertTrue(compact.contains("50.25"))
    }

    @Test
    fun testDetailedFormatting() {
        val amount = BigDecimal("75.50")

        val detailed = CurrencyFormatter.formatDetailed(amount, Currency.GBP)
        assertTrue(detailed.contains("£"))
        assertTrue(detailed.contains("GBP"))
    }

    @Test
    fun testExtensionFunctions() {
        val amount = BigDecimal("100.00")

        val formatted = amount.formatted(Currency.USD)
        assertTrue(formatted.contains("$"))

        val compact = amount.formattedCompact(Currency.AED)
        assertTrue(compact.contains("د.إ"))

        val detailed = amount.formattedDetailed(Currency.EUR)
        assertTrue(detailed.contains("€"))
        assertTrue(detailed.contains("EUR"))
    }

    @Test
    fun testCurrencyParsing() {
        // Test parsing formatted currency
        val parsed = CurrencyFormatter.parse("$100.50", Currency.USD)
        assertEquals(BigDecimal("100.50"), parsed)

        // Test parsing with different symbols
        val parsedAED = CurrencyFormatter.parse("د.إ 250.75", Currency.AED)
        assertEquals(BigDecimal("250.75"), parsedAED)

        // Test parsing just numbers
        val parsedPlain = CurrencyFormatter.parse("50.25", Currency.USD)
        assertEquals(BigDecimal("50.25"), parsedPlain)
    }

    // MARK: - Voice Currency Detector Tests

    @Test
    fun testSimpleAmountCurrencyExtraction() {
        // Test "50 dollars"
        val result1 = VoiceCurrencyDetector.extractAmountAndCurrency("50 dollars")
        assertNotNull(result1)
        assertEquals(50.0, result1!!.first, 0.01)
        assertEquals(Currency.USD, result1.second)

        // Test "د.إ 100"
        val result2 = VoiceCurrencyDetector.extractAmountAndCurrency("د.إ 100")
        assertNotNull(result2)
        assertEquals(100.0, result2!!.first, 0.01)
        assertEquals(Currency.AED, result2.second)
    }

    @Test
    fun testComplexVoiceCommandExtraction() {
        // Test full sentence
        val result1 = VoiceCurrencyDetector.extractAmountAndCurrency(
            "I just spent 25.50 euros on groceries"
        )
        assertNotNull(result1)
        assertEquals(25.50, result1!!.first, 0.01)
        assertEquals(Currency.EUR, result1.second)

        // Test with merchant
        val result2 = VoiceCurrencyDetector.extractAmountAndCurrency(
            "Paid 20 pounds at the store"
        )
        assertNotNull(result2)
        assertEquals(20.0, result2!!.first, 0.01)
        assertEquals(Currency.GBP, result2.second)
    }

    @Test
    fun testCurrencyDetectionWithoutAmount() {
        val usd = VoiceCurrencyDetector.detectCurrency(
            text = "spent some dollars",
            defaultCurrency = Currency.AED
        )
        assertEquals(Currency.USD, usd)

        val aed = VoiceCurrencyDetector.detectCurrency(
            text = "paid in dirhams",
            defaultCurrency = Currency.USD
        )
        assertEquals(Currency.AED, aed)

        // Test default fallback
        val defaultCurrency = VoiceCurrencyDetector.detectCurrency(
            text = "spent money",
            defaultCurrency = Currency.GBP
        )
        assertEquals(Currency.GBP, defaultCurrency)
    }

    @Test
    fun testContainsCurrency() {
        assertTrue(VoiceCurrencyDetector.containsCurrency("50 dollars"))
        assertTrue(VoiceCurrencyDetector.containsCurrency("د.إ 100"))
        assertTrue(VoiceCurrencyDetector.containsCurrency("paid in euros"))
        assertFalse(VoiceCurrencyDetector.containsCurrency("spent money"))
    }

    @Test
    fun testNormalizeCurrencySymbols() {
        val normalized = VoiceCurrencyDetector.normalizeCurrencySymbols(
            "$50 and €25 and د.إ 100"
        )
        assertTrue(normalized.contains("USD"))
        assertTrue(normalized.contains("EUR"))
        assertTrue(normalized.contains("AED"))
    }

    // MARK: - Integration Tests

    @Test
    fun testEndToEndCurrencyFlow() {
        // Simulate voice command: "I spent 50 dirhams on groceries"
        val voiceCommand = "I spent 50 dirhams on groceries"

        // 1. Detect currency
        val currency = Currency.detectFromText(voiceCommand)
        assertNotNull(currency)
        assertEquals(Currency.AED, currency)

        // 2. Extract amount
        val extraction = VoiceCurrencyDetector.extractAmountAndCurrency(voiceCommand)
        assertNotNull(extraction)
        assertEquals(50.0, extraction!!.first, 0.01)
        assertEquals(Currency.AED, extraction.second)

        // 3. Format for display
        val formatted = CurrencyFormatter.formatCompact(
            BigDecimal(extraction.first.toString()),
            extraction.second
        )
        assertTrue(formatted.contains("د.إ"))
        assertTrue(formatted.contains("50"))
    }
}
